### ISSUE TITLE
Quick pick focused item does not meet color contrast of 3:1

### ISSUE THREAD

**Issue by ejizba** [microsoft/vscode#121493]
_originally filed in https://github.com/microsoft/vscode-azurefunctions/issues/2680_

Issue Type: <b>Bug</b>

The color contrast of the focused item in a quick pick compared to the non-focused item is only 1.1:1, aka less than the required 3:1
![Screen Shot 2021-04-16 at 11 18 40 AM](https://user-images.githubusercontent.com/11282622/115067775-3a038100-9ea6-11eb-8e58-4ccf1e4cf99e.png)
<img width="277" alt="Screen Shot 2021-04-16 at 11 23 03 AM" src="https://user-images.githubusercontent.com/11282622/115067868-57384f80-9ea6-11eb-8b25-9f0b51971a93.png">

Might I suggest you match the pattern you use in the file explorer, which _does_ meet the 3:1 contrast ratio (either the filled version or the border-only version)
![Screen Shot 2021-04-16 at 11 19 37 AM](https://user-images.githubusercontent.com/11282622/115068602-505e0c80-9ea7-11eb-941b-58e0a4141029.png)![Screen Shot 2021-04-16 at 11 19 51 AM](https://user-images.githubusercontent.com/11282622/115068603-50f6a300-9ea7-11eb-9e05-97c949c4ddb3.png)

VS Code version: Code - Insiders 1.56.0-insider (394a2e4779ff767e2761a59a51b15848dcd24173, 2021-04-13T11:11:34.416Z)
OS version: Darwin x64 20.3.0

<details>
<summary>System Info</summary>

|Item|Value|
|---|---|
|CPUs|Intel(R) Core(TM) i7-8850H CPU @ 2.60GHz (12 x 2600)|
|GPU Status|2d_canvas: enabled<br>gpu_compositing: enabled<br>metal: disabled_off<br>multiple_raster_threads: enabled_on<br>oop_rasterization: enabled<br>opengl: enabled_on<br>protected_video_decode: unavailable_off<br>rasterization: enabled<br>skia_renderer: disabled_off_ok<br>video_decode: enabled<br>webgl: enabled<br>webgl2: enabled|
|Load (avg)|2, 2, 2|
|Memory (System)|16.00GB (0.51GB free)|
|Process Argv|--crash-reporter-id 63154ab5-ba0e-419a-aa36-bd67d86db27f|
|Screen Reader|no|
|VM|0%|
</details>

**Comment by misolori** [comment]
This seems like a bug with whoever took the color contrast colors because i'm not seeing the same colors. We don't ever use `#D6EBFF` or `#f3f3f3` as a foreground color. Here are the color contrast ratio for default and selected items:

![image](https://user-images.githubusercontent.com/35271042/115270185-edf84c80-a0f0-11eb-976e-a36674950388.png)

![image](https://user-images.githubusercontent.com/35271042/115270171-ec2e8900-a0f0-11eb-976e-4d9100b62103.png)

**Comment by ejizba** [comment]
@misolori The accessibility rule I'm talking about applies to **_not-text_** content, which is why I never reference your foreground color "#616161". https://www.w3.org/WAI/WCAG21/Understanding/non-text-contrast.html

> The visual presentation of the following have a contrast ratio of at least 3:1 against adjacent color(s):
> Visual information required to identify user interface components and **states**, except for inactive components or where the appearance of the component is determined by the user agent and not modified by the author;

People need to know which item is selected and they may have a difficult time knowing because the ratio of #D6EBFF to #F3F3F3 does not meet 3:1.

This was obviously already fixed for the file explorer (where it's much easier to see which one is selected) - is there some reason you're opposed to using those colors?

**Comment by misolori** [comment]
I'm not sure i'm understanding the request. My understanding is that the non-text contrast rule applies to things like icons and not background colors that are next to each other. That is why all of the contrast tools ask for a foreground and a background color.

Additionally, when in the quick pick the focus is on the input box at the top and selection is below, so the interaction is different:

![image](https://user-images.githubusercontent.com/35271042/116146627-8cf1eb00-a693-11eb-8e5a-6818e00c43b2.png)

cc @isidorn @joaomoreno for more input

**Comment by ejizba** [comment]
They're not _just_ "background colors that are next to each other", though. The key point is that the background color is the only thing that conveys the state (selected vs. non-selected).

If you ignore the nitty gritty of the rule for a second and just focus on the described intent, the quick pick is useless if you can't tell the difference between the background color of "Tasks: Run Task" and "View: Reopen Editor With..." - you'd never know which command you're about to run.

**Comment by misolori** [comment]
Looping in other people because this has broader impact. I think what you are referring to is a color meaning and not a color contrast issue:

[Understanding Success Criterion 1.4.1: Use of Color](https://www.w3.org/WAI/WCAG21/Understanding/use-of-color.html)

> Color is not used as the only visual means of conveying information, indicating an action, prompting a response, or distinguishing a visual element.

However, we do have additional indicators for when an item is selected in the quick pick (we also show a gear):

![image](https://user-images.githubusercontent.com/35271042/116158721-009bf400-a6a4-11eb-9e47-7d0f1c678285.png)

**Comment by ejizba** [comment]
Yeah I would say both "Use of Color" and "Non-text contrast" probably apply. Both mention the "3:1" ratio in a few places, so I think the fix is the same.

> However, we do have additional indicators for when an item is selected in the quick pick (we also show a gear)

The issue is generic to all quick picks and most do not have the gear.

**Comment by isidorn** [comment]
I think this request makes sense. In short:
Quick Pick should follow the new List UI to show the focused element with a strong border - like we do in the explorer, and not with the background color. I think that might work great. Once we try it out if we are not happy we can change the background color.
Thus unassigning @joaomoreno and assigning @TylerLeonhardt

**Comment by misolori** [comment]
@isidorn i think the tricky part here is focus is on the input in the quick pick, so you can't really apply the border to the selection. We could use the strong blue bg color but I also don't know why that wasn't used in the beginning so wanted to know more context there.

**Comment by isidorn** [comment]
@misolori my best guess is that in the beginning we used the light blue since it was the default selection color.
I would suggest to change to the strong blue bg color start of milestone and we see how it feels - we will get lots of feedback.

fyi @bpasero since he might be the original author.

**Comment by bpasero** [comment]
I think the explorer color and the quick pick color have gone different directions, mainly I guess because the background color is different:

![image](https://user-images.githubusercontent.com/900690/116245766-92216980-a769-11eb-8ba9-bde344c6af67.png)

One special thing of quick pick is that the list typically never has keyboard focus but only the input field above.

**Comment by misolori** [comment]
So the suggestion would be to match the explorer for active selection. This would apply to both dark and light themes:

![image](https://user-images.githubusercontent.com/35271042/116265300-dc85f480-a72f-11eb-96ee-bd606e0d6e36.png)

![image](https://user-images.githubusercontent.com/35271042/116265314-e14aa880-a72f-11eb-999a-bb323f4e27e3.png)

This is not easily supported and requires more theming work (keybindings, actions, filter, etc.), so I'll wait for @TylerLeonhardt to respond when he gets back.

**Comment by isidorn** [comment]
Looks great in dark theme 😎

**Comment by bpasero** [comment]
I don't mind as long as I can tweak that color in my settings as I prefer the current colors.

**Comment by bpasero** [comment]
@misolori are we missing a color for the highlight text foreground when an item is selected in suggest:

![image](https://user-images.githubusercontent.com/900690/117249816-835e4500-ae42-11eb-816f-e675196c7f26.png)

**Comment by bpasero** [comment]
Same question in quick pick:

![image](https://user-images.githubusercontent.com/900690/117250706-e8666a80-ae43-11eb-93cf-845a33bbcd6d.png)

**Comment by misolori** [comment]
@bpasero have you tried `list.focusHighlightForeground`?

**Comment by bpasero** [comment]
@misolori yeah that works, I think there is still a regression though, e.g. references view:

![image](https://user-images.githubusercontent.com/900690/117258542-5f543100-ae4d-11eb-852b-d8fe457b8d0f.png)

**Comment by usernamehw** [comment]
Is it intentional that active icon has different color now? It looks worse than in Stable.

![active_suggest_item](https://user-images.githubusercontent.com/9638156/119040449-1c05d080-b9be-11eb-917e-92e0495bfd28.png)

**Comment by misolori** [comment]
@usernamehw this is one of the downsides when themes deviate from the default style for active items. The default is to have the active item have the highest contrast so that the foreground color is inverted:

![image](https://user-images.githubusercontent.com/35271042/119047072-4a5dbe80-b972-11eb-96a9-281a80914f1a.png)


We have the same problem in all list views as well where there are colored icons:

![image](https://user-images.githubusercontent.com/35271042/119047785-349cc900-b973-11eb-8ef7-fe520caf2412.png)

![image](https://user-images.githubusercontent.com/35271042/119047921-62820d80-b973-11eb-8325-cf5e6ef45218.png)

![image](https://user-images.githubusercontent.com/35271042/119048783-8f82f000-b974-11eb-9bcf-99f3f3a375a7.png)

**Comment by misolori** [comment]
@ejizba can you please verify that this has now been fixed? Thanks.

**Comment by misolori** [comment]
Verification steps:
- Open the quick pick in our default dark and light, verify that the focused item bg & text pass the cc ratio of 4.5:1 ([you can use this tool](https://www.tpgi.com/color-contrast-checker/))
- Test that the following theme colors work:
    -  `list.focusHighlightForeground` highlight color for focused item in any list
    - `quickInputList.focusForeground` foreground color for focused item in quick pick
    - `editorSuggestWidget.selectedForeground` foreground color for selected item in suggest widget

**Comment by rzhao271** [comment]
@misolori do we have to check the contrast between the background colours of the selected item and the unselected ones for the quick pick in dark mode? If so, that ratio is currently below 3:1. Things look fine in the default light mode, though.

**Comment by misolori** [comment]
@rzhao271 contrast is just between background color and foreground text

**Comment by ejizba** [comment]
> contrast is just between background color and foreground text

That is not true - the point of the issue was "the contrast between the background colours of the selected item and the unselected ones" just as @rzhao271 was asking about. This is based on the [non-text content rule](https://www.w3.org/WAI/WCAG21/Understanding/non-text-contrast.html) which says:

> any visual information necessary to indicate state, such as whether a component is selected or focused must also ensure that the information used to identify the control in that state has a minimum 3:1 contrast ratio.
